### PR TITLE
hyper-v: Remove snapshots and not leak vhds

### PR DESCRIPTION
**Pull request by ader1990**
If the snapshots are not removed, the VHDs will be left on the disk after vm deletion, as the *.avhdx are the only one removed.

On localhost, do not use invoke command with computername set, as it might fail on non domain computers.

### CODE REVIEW

File: Libraries/HyperV.psm1 [hunk 1]
```diff
@@ -258,13 +258,35 @@ Function DeleteHyperVGroup([string]$HyperVGroupName, [string]$HyperVHost) {
         LogMsg "Stop-VM -Name $($_.Name) -Force -TurnOff "
         $vm = $_
         Stop-VM -Name $vm.Name -Force -TurnOff -ComputerName $HyperVHost
+        Remove-VMSnapshot -VMName $vm.Name -ComputerName $HyperVHost -Confirm:$false
+        if ($?) {
```
[the remark this answers is not in the document]
Comment (ader1990) [reply Libraries/HyperV.psm1]: You are correct, I will make the required changes, thank you.

File: Libraries/HyperV.psm1 [hunk 2]
```diff
@@ -742,6 +768,27 @@ function Wait-VMState {
...
+        $VMName,
+        $VMStatus,
+        $HvServer,
+        $RetryCount=30,
+        $RetryInterval=5
+    )
+
+    $currentRetryCount = 0
+    while ($currentRetryCount -lt $RetryCount -and `
+              (Get-VM -ComputerName $hvServer -Name $vmName).Status -ne $VMStatus) {
```
Comment (adriansuhov) [inline Libraries/HyperV.psm1]: $HvServer instead of $hvServer, $VMName instead of $vmName

File: Libraries/HyperV.psm1 [hunk 3]
```diff
@@ -258,13 +258,39 @@ Function DeleteHyperVGroup([string]$HyperVGroupName, [string]$HyperVHost) {
...
         Stop-VM -Name $vm.Name -Force -TurnOff -ComputerName $HyperVHost
+        Remove-VMSnapshot -VMName $vm.Name -ComputerName $HyperVHost `
+            -IncludeAllChildCheckpoints -Confirm:$false
+        if (!$?) {
+            LogErr ("Failed to remove snapshots for VM {0}" -f @($vm.Name))
+            return $false
+        }
+        Wait-VMStatus -VMName $vm.Name -VMStatus "Operating Normally" -RetryInterval 2 `
+            -HvServer $HyperVHost
+        $vm = Get-VM -Name $vm.Name
```
Comment (adriansuhov) [inline Libraries/HyperV.psm1]: $vm = Get-VM -Name $vm.Name **-ComputerName $HyperVHost**

File: Libraries/HyperV.psm1 [hunk 4]
```diff
@@ -258,13 +258,39 @@ Function DeleteHyperVGroup([string]$HyperVGroupName, [string]$HyperVHost) {
...
+            }
+            if ($HyperVHost -ne "localhost" -and $HyperVHost -ne $(hostname)) {
+                $invokeCommandParams.ComputerName = $HyperVHost
+            }
+            Invoke-Command @invokeCommandParams
+            if (!$?) {
+                LogMsg "Failed to remove ${vhdPath} using Invoke-Command"
+                $vhdUncPath = $vhdPath -replace '^(.):', "\\$(HyperVHost)\`$1$"
+                LogMsg "Removing ${vhdUncPath} ..."
+                Remove-Item -Path $vhdUncPath -Force -ErrorAction "Stop"
```
[the remark this answers is not in the document]
Comment (ader1990) [reply Libraries/HyperV.psm1]: I checked and I need to remove the stop flag

**Pull request merged**